trunner/device.py

```diff
@@ -30,13 +33,203 @@
+        self.baudrate = baudrate
+        self.dir = dir
+        self.cwd = cwd
+        self.proc = None
+        self.reader_thread = None
+        self.reader_lock = None
+        self.wait_dispatcher = wait_dispatcher
+        self.output_buffer = ''
+
+    def _reader(self):
+        while True:
+            line = self.proc.readline()
+            if not line:
+                break
+
+            if self.wait_dispatcher:
+                msg = f'Starting message dispatcher on [{self.port}] (speed={self.baudrate})'
+                if msg in line:
+                    self.wait_dispatcher = False
+                    self.reader_lock.release()
+
+            self.output_buffer += line
+
+    def run(self):
+        # Use pexpect.spawn to run process as PTY, so it will flush at new line
+        self.proc = pexpect.spawn(
+            'phoenixd',
+            ['-p', self.port,
+             '-b', str(self.baudrate),
+             '-s', self.dir],
+            cwd=self.cwd,
+            encoding='utf-8'
+        )
+
+        if self.wait_dispatcher:
+            self.reader_lock = threading.Lock()
+            self.reader_lock.acquire()
+
+        self.reader_thread = threading.Thread(target=self._reader)
+        self.reader_thread.start()
+
+        if self.wait_dispatcher:
+            # When dispatcher starts then reader will release lock
+            self.reader_lock.acquire(timeout=10)
+
+        return self.proc
+
+    def output(self):
+        output = self.output_buffer
+        if is_github_actions():
+            output = '::group::Pheonixd\n' + output + '::endgroup::\n'
+
+        return output
+
+    def kill(self):
+        os.killpg(os.getpgid(self.proc.pid), signal.SIGTERM)
+        self.reader_thread.join()
+        if self.reader_lock.locked():
+            self.reader_lock.release()
+
+    def __enter__(self):
+        self.run()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.kill()
+
+
+class PloError(Exception):
+    def __init__(self, message, expected):
+        msg = Color.colorify("PLO ERROR:\n", Color.BOLD)
+        msg += str(message) + '\n'
+        if expected:
+            msg += Color.colorify("EXPECTED:\n", Color.BOLD)
+            msg += str(expected) + '\n'
+
+        super().__init__(msg)
+
+
+class PloTalker:
+    """Interface to communicate with plo"""
+
+    def __init__(self, port, baudrate=115200):
+        self.port = port
+        self.baudrate = baudrate
+        self.serial = None
+        self.plo = None
+
+    def open(self):
+        try:
+            self.serial = serial.Serial(self.port, baudrate=self.baudrate)
+        except serial.SerialException:
+            logging.error(f'Port {self.port} not available\n')
+            raise
+
+        self.plo = pexpect.fdpexpect.fdspawn(self.serial, timeout=8)
+        return self
+
+    def close(self):
+        self.serial.close()
+
+    def __enter__(self):
+        return self.open()
```

Comment:
See comment about Phoenixd.\__enter\__

Reply:
I added catching exception from `pexpect.fdspawn` and closing serial port if exception occures.